### ISSUE TITLE
Kubernetes manifest questions

### ISSUE THREAD

**Issue by weeco** [grafana/agent#108]
I stumbled across the deployment manifest here: https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml and I have a few questions:

1. You are using `minReadySeconds` (https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml#L250), I assume because there are no liveness/readiness probes (yet). Could I use the `/metrics` endpoint instead or what's the purpose behind using `minReadySeconds`?

2. Security context is set to root user / privileged (https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml#L274-L276). Why does the agent need root permissions? Opening a socket on port 80 is probably one reason, but I think the port could easily be changed to another port which does not require root permissions?

3. There are two different paths specified for the `wal_directory`. Does the argument overwrite the setting from the YAML config or are these two different config options? See: https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml#L203 and https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml#L262

**Comment by rfratto** [comment]
Hi, these are all good questions and I'd like to provide some extra context about why that manifest has some (glaring) problems like this: the Kubernetes manifest you've found is generated from our Tanka configs, which was hastily slapped together without too much thought for the initial v0.1.0 release. It's mostly compromised of different parts of the prometheus-ksonnet and Grafana Loki's Promtail's Tanka configs. 

> You are using minReadySeconds (https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml#L250), I assume because there are no liveness/readiness probes (yet). Could I use the /metrics endpoint instead or what's the purpose behind using minReadySeconds?

This is just the default that our Tanka libraries apply to replica sets and deployments. Here's the [reasoning](https://github.com/grafana/jsonnet-libs/blob/3c072927d359ffa70d4a066033da90f8668c41c1/ksonnet-util/kausal.libsonnet#L143-L144) from the code: 

```jsonnet
// We want to specify a minReadySeconds on every deamonset, so we get some
// very basic canarying, for instance, with bad arguments.
```

> Security context is set to root user / privileged (https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml#L274-L276). Why does the agent need root permissions? Opening a socket on port 80 is probably one reason, but I think the port could easily be changed to another port which does not require root permissions?

Good point, it doesn't need root permissions for anything other than listening on port 80 by default. It makes a ton of sense to change this. 

> There are two different paths specified for the wal_directory. Does the argument overwrite the setting from the YAML config or are these two different config options? See: https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml#L203 and https://github.com/grafana/agent/blob/master/production/kubernetes/agent.yaml#L262

The flag overrides the value from YAML, so we definitely don't need both. IMO it makes more sense to remove the flag and point the YAML at the correct folder instead (just so all configuration is kept to being within the YAML). 

Hope this helps, let me know if there's anything else I can help with!

**Comment by weeco** [comment]
Thank you for clarifying. Indeed I have 1-2 more questions after further browsing through the documentation. 

```
# Whether this agent instance should only scrape from targets running on the
# same machine as the agent process.
[host_filter: <boolean> | default = false]

# How frequently the WAL should be truncated of its old data.
[wal_truncate_frequency: <duration> | default = "1m"]

# Deadline for flushing data to the remote write before marking it
# as an error.
[remote_flush_deadline: <duration> | default = "1m"]
```

1. If I understand `host_filter` correctly it makes sense to have this enabled if `agent` is deployed as daemonset. Otherwise all the `agents` would scrape metrics from other nodes as well and therefore scrape the same metric series/targets multiple times which is probably something you want to avoid? In the examples agent is deployed as daemonset and `host_filter` is disabled, why?

2. I have difficulties understanding the options `wal_truncate_frequency` and `remote_flush_deadline`.  Would setting `wal_truncate_frequency` to 1m (default) mean if my remote write target is down/unreachable for longer than 1min that `agent` would not be able to backfill the metrics because it truncates all the data older than 1min? Does `remote_flush_deadline` in practice mean that `agent` will stop retrying/sending data to my remote write target if the data becomes older than 1minute?

```
# Whether this agent instance should only scrape from targets running on the
# same machine as the agent process.
[host_filter: <boolean> | default = false]

# How frequently the WAL should be truncated of its old data.
[wal_truncate_frequency: <duration> | default = "1m"]

# Deadline for flushing data to the remote write before marking it
# as an error.
[remote_flush_deadline: <duration> | default = "1m"]
```

**Comment by rfratto** [comment]
> In the examples agent is deployed as daemonset and host_filter is disabled, why?

Ah, this is actually a bug introduced recently; I just opened #109 to fix this. Your understanding of `host_filter` is right; it must be `true` when deploying as a DaemonSet and `false` when deployed as a statefulset/deployment.

> I have difficulties understanding the options wal_truncate_frequency and remote_flush_deadline.

This is an area where the documentation could use an improvement. The `wal_truncate_frequency` is the period where we create a new WAL checkpoint and only delete data whose timestamp is older than the last successful `remote_write` timestamp. It also creates a new WAL segment for new appends. 

> Does remote_flush_deadline in practice mean that agent will stop retrying/sending data to my remote write target if the data becomes older than 1minute?

I looked through Prometheus again and it seems like I misunderstood what `remote_flush_deadline` is used for. `remote_flush_deadline` is the deadline to wait for flushes when the remote_writer is shutting down. For the agent, this comes into play when:

1. The Agent process shuts down 
2. Using the scraping service and a config gets resharded to a new node or updated on the same node 

I'll update the reference to make note of these two points, sorry about the confusion here.
